File: src/main/java/com/saomc/saoui/SoundCore.java
package com.saomc.saoui;

import com.saomc.saoui.config.OptionCore;
import net.minecraft.client.Minecraft;
import net.minecraft.client.audio.PositionedSoundRecord;
import net.minecraft.client.audio.SoundHandler;
import net.minecraft.entity.Entity;
import net.minecraft.util.ResourceLocation;

public class SoundCore {

    public static final String CONFIRM = "sao.confirm";
    public static final String DIALOG_CLOSE = "sao.dialog.close";
    public static final String MENU_POPUP = "sao.menu.popup";
    public static final String MESSAGE = "sao.message";
    public static final String ORB_DROPDOWN = "sao.orb.dropdown";
    public static final String PARTICLES_DEATH = "sao.particles.death";
    public static final String LOW_HEALTH = "sao.low.health";

    private static ResourceLocation getResource(String name) {
        return new ResourceLocation(SAOCore.MODID, name);
    }

    public static boolean isSfxPlaying(String name) {
        return Minecraft.getMinecraft().getSoundHandler().isSoundPlaying(create(getResource(name)));
    }

    public static void playFromEntity(Entity entity, String name) {
        if (entity != null) {
            playAtEntity(entity, name);
        }
    }

    public static void playAtEntity(Entity entity, String name) {
        final Minecraft mc = Minecraft.getMinecraft();

        if (mc.theWorld != null && mc.theWorld.isRemote) {
            play(mc.getSoundHandler(), name, (float) entity.posX, (float) entity.posY, (float) entity.posZ);
        }
    }

    public static void play(Minecraft mc, String name) {
        if (mc != null && mc.theWorld != null && mc.theWorld.isRemote) {
            play(mc.getSoundHandler(), name);
        }
    }

    public static void play(SoundHandler handler, String name) {
        if ((OptionCore.SOUND_EFFECTS.isEnabled()) && (handler != null)) {
            handler.playSound(create(getResource(name)));
        }
    }

    private static void play(SoundHandler handler, String name, float x, float y, float z) {
        if ((OptionCore.SOUND_EFFECTS.isEnabled()) && (handler != null)) {
            handler.playSound(create(getResource(name), x, y, z));
        }
    }

    public static void resumeSounds(Minecraft mc) {
        mc.getSoundHandler().resumeSounds();
    }

    //Helper functions - 1.8.8 mirror
    private static PositionedSoundRecord create(ResourceLocation soundResource) {
        return new PositionedSoundRecord(soundResource, 1.0F, 1.0F, 0.0F, 0.0F, 0.0F);
    }

    private static PositionedSoundRecord create(ResourceLocation soundResource, float xPosition, float yPosition, float zPosition) {
        return new PositionedSoundRecord(soundResource, 4.0F, 1.0F, xPosition, yPosition, zPosition);
    }
}

File: src/main/java/com/saomc/saoui/SoundCore.kt
/*
 * Copyright (C) 2016-2019 Arnaud 'Bluexin' Solé
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.saomc.saoui

import com.teamwizardry.librarianlib.features.kotlin.Minecraft
import net.minecraft.entity.Entity
import net.minecraft.util.ResourceLocation
import net.minecraft.util.SoundCategory
import net.minecraft.util.SoundEvent
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

enum class SoundCore {

    CONFIRM,
    DIALOG_CLOSE,
    MENU_POPUP,
    MESSAGE,
    ORB_DROPDOWN,
    PARTICLES_DEATH;

    val sound = SoundEvent(ResourceLocation(SAOCore.MODID, name.toLowerCase()))

    companion object {
        @SubscribeEvent
        fun registerSoundEvent(event: RegistryEvent.Register<SoundEvent>){
            values().forEach { event.registry.register(it.sound.setRegistryName(it.name.toLowerCase())) }
        }
    }
}

fun SoundCore.play() = Minecraft().player.playSound(this.sound, 1.0f, 1.0f)
fun SoundCore.playAtEntity(ent: Entity) = ent.world.playSound(ent.posX, ent.posY, ent.posZ, this.sound, SoundCategory.AMBIENT, 1.0f, 1.0f, true)